### PR TITLE
[control-plane-manager] added ability to add  kube-scheduler extenders

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
## Description

Added ability to add kube-scheduler extenders.

## Why do we need it, and what problem does it solve?

We need to use extensions for kube-scheduler in projects like sds and virtualization.

* In order to place pods as close to data as possible in case of using sds-replicated-volume. This allows you to improve the performance of the volumes. 
* To define zones in virtualization.
* To prioritize nodes with large disk size in case of using sds-local-volume.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: control-plane
type: feature
summary: Added ability to connect  kube-scheduler extenders.
impact: kube-scheduler should be restarted every time when extender config is added.
impact_level: default
```

### CODE REVIEW

**Comment by RomanenkoDenys** [comment]
kube-scheduler config without extenders.

```
root@dev-master-0:~# cat /etc/kubernetes/deckhouse/extra-files/scheduler-config.yaml

apiVersion: kubescheduler.config.k8s.io/v1
kind: KubeSchedulerConfiguration
clientConnection:
  kubeconfig: /etc/kubernetes/scheduler.conf
profiles:
- pluginConfig:
  - name: PodTopologySpread
    args:
      defaultingType: List
      defaultConstraints:
      - maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: ScheduleAnyway
```

Add extender:
```
root@dev-master-0:~# cat extender.yaml
apiVersion: deckhouse.io/v1alpha1
kind: KubeSchedulerWebhookConfiguration
metadata:
  name: sds-replicated-volume
webhooks:
- weight: 5
  failurePolicy: Ignore
  clientConfig:
    service:
      name: scheduler
      namespace: d8-sds-replicated-volume
      port: 8080
      path: /scheduler
    caBundle: ABCD=
  timeoutSeconds: 5
```

Kube-scheduler config after extender add:
```
cat /etc/kubernetes/deckhouse/extra-files/scheduler-config.yaml

apiVersion: kubescheduler.config.k8s.io/v1
kind: KubeSchedulerConfiguration
clientConnection:
  kubeconfig: /etc/kubernetes/scheduler.conf
profiles:
- pluginConfig:
  - name: PodTopologySpread
    args:
      defaultingType: List
      defaultConstraints:
      - maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: ScheduleAnyway
extenders:
- urlPrefix: https://scheduler.d8-sds-replicated-volume.cluster.local:8080/scheduler
  filterVerb: filter
  prioritizeVerb: prioritize
  weight: 5
  enableHTTPS: true
  httpTimeout: 5s
  nodeCacheCapable: true
  ignorable: true
  tlsConfig:
    caData: ABCD=
```

**Comment by RomanenkoDenys** [comment]
```
kubectl get kubeschedulerwebhookconfiguration.deckhouse.io

NAME                    NAME        NAMESPACE                  PORT   PATH         TIMEOUT   WEIGHT   FAILURE POLICY
sds-replicated-volume   scheduler   d8-sds-replicated-volume   8080   /scheduler   5         5        Ignore
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          required: ['webhooks']
+          description: |
+            Определяет конфигурацию для включения внешних extender вебхуков для kube-scheduler.
+
+            С их помощью можно гибко настраивать кластер Kubernetes.
```
Comment (name212) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:27]:
```suggestion
            С их помощью можно гибко настраивать планировщик кластера Kubernetes.
```

**Review by name212 — APPROVED** [review]

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  versions:
+    - name: v1alpha1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          required: ['webhooks']
+          description: |
+            Определяет конфигурацию для включения внешних extender вебхуков для kube-scheduler.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:25]:
```suggestion
            Определяет конфигурацию для включения внешних вебхуков kube-scheduler.
```

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+              items:
+                type: object
+                required: ['weight', 'clientConfig']
+                properties:
+                  weight:
+                    type: integer
+                    description: |
+                      Числовой множитель для оценок узла, которые генерирует вызов расчета приоритезации.
+
+                      Вес должен быть положительным целым числом.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:40]:
```suggestion
                      Числовой множитель для оценок узла, которые генерирует вызов расчета приоритезации.
```
Типа параметра уже указан

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    minimum: 1
+                  failurePolicy:
+                    type: string
+                    enum: ['Fail', 'Ignore']
+                    description: |
+                      Определяет политику - падать с ошибкой или пропускать если extender вернул ошибку или недоступен.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:46]:
```suggestion
                      Политика реакции на событие, если вебхук вернул ошибку или недоступен:
                      `Fail` — падать с ошибкой;
                      `Ignore` — пропускать (игнорировать).
```

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    default: Fail
+                  clientConfig:
+                    type: object
+                    required: ['service', 'caBundle']
+                    description: |
+                      Конфигурация webhook.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:52]:
```suggestion
                      Конфигурация вебхука.
```

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                            description: |
+                              Порт сервиса.
+                          path:
+                            type: string
+                            description: |
+                              URI путь.
+                      caBundle:
+                        type: string
+                        description: |
+                          CA bundle для проверки сертификата webhook.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:79]:
```suggestion
                          CA bundle для проверки сертификата вебхука.
```

File: modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                  timeoutSeconds:
+                    type: integer
+                    description: |
+                      Таймаут срабатывания webhook.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/doc-ru-kube_scheduler_webhook_configuration.yaml:83]:
```suggestion
                      Таймаут срабатывания вебхука.
```

File: modules/040-control-plane-manager/crds/kube_scheduler_webhook_configuration.yaml [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+              items:
+                type: object
+                required: ['weight', 'clientConfig']
+                properties:
+                  weight:
+                    type: integer
+                    description: |
+                      The numeric multiplier for the node scores that the prioritize call generates.
+
+                      The weight should be a positive integer.
```
Comment (z9r5) [inline modules/040-control-plane-manager/crds/kube_scheduler_webhook_configuration.yaml:40]:
```suggestion
                      The numeric multiplier for the node scores that the prioritize call generates.
```

File: modules/040-control-plane-manager/docs/EXAMPLES.md [hunk 9]
```diff
@@ -19,3 +21,23 @@ spec:
...
+- weight: 5
+  failurePolicy: Ignore
+  clientConfig:
+  service:
+  name: scheduler
+  namespace: d8-sds-replicated-volume
+  port: 8080
+  path: /scheduler
+  caBundle: ABCD=
+  timeoutSeconds: 5
```
Comment (AleksZimin) [inline modules/040-control-plane-manager/docs/EXAMPLES.md:42]:
```suggestion
apiVersion: deckhouse.io/v1alpha1
 kind: KubeSchedulerWebhookConfiguration
 metadata:
   name: sds-replicated-volume
 webhooks:
 - weight: 5
   failurePolicy: Ignore
   clientConfig:
     service:
       name: scheduler
       namespace: d8-sds-replicated-volume
       port: 8080
       path: /scheduler
     caBundle: ABCD=
   timeoutSeconds: 5
```

File: modules/040-control-plane-manager/hooks/kube_scheduler_extenders.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+		Ignorable bool   `yaml:"ignorable" json:"ignorable"`
+		CAData    string `yaml:"caData" json:"caData"`
+	}
+	extenders := make([]extenderConfig, 0)
+
+	var clusterDomain = input.Values.Get("global.discovery.clusterDomain").String()
+
+	for _, snapshot := range input.Snapshots["kube_scheduler_extenders"] {
+		for _, config := range snapshot.([]KubeSchedulerWebhook) {
+			urlPrefix, err := url.JoinPath(fmt.Sprintf("https://%s.%s.%s:%d", config.ClientConfig.Service.Name, config.ClientConfig.Service.Namespace, clusterDomain, config.ClientConfig.Service.Port), config.ClientConfig.Service.Path)
```
Comment (AleksZimin) [inline modules/040-control-plane-manager/hooks/kube_scheduler_extenders.go:67]:
```suggestion
			urlPrefix, err := url.JoinPath(fmt.Sprintf("https://%s.%s.svc.%s:%d", config.ClientConfig.Service.Name, config.ClientConfig.Service.Namespace, clusterDomain, config.ClientConfig.Service.Port), config.ClientConfig.Service.Path)
```

**Review by z9r5 — APPROVED** [review]

**Comment by z9r5** [comment]
Made some updates to the documentation.

File: modules/040-control-plane-manager/docs/EXAMPLES.md (hunk 9 above)
Comment (z9r5) [reply modules/040-control-plane-manager/docs/EXAMPLES.md]: Fixed

File: modules/040-control-plane-manager/hooks/kube_scheduler_extenders.go (hunk 10 above)
Comment (RomanenkoDenys) [reply modules/040-control-plane-manager/hooks/kube_scheduler_extenders.go:67]: thanks!

**Review by nabokihms — APPROVED** [review]

**Pull request merged**